eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentLifecycle.scala

```diff
@@ -0,0 +1,330 @@
+      case ((MilliSatoshi(0), payments), _) => (0 msat, payments)
+      case ((remaining, payments), channel) =>
+        val childPayment = createChildPayment(nodeParams, request, remaining.min(channel.canSend), channel)
+        (remaining - childPayment.finalPayload.amount, childPayment +: payments)
+    }
+    // For other channels, we need to split the amount based on network statistics and pessimistic fees estimates.
+    val maxFeePct = request.routeParams.map(_.maxFeePct).getOrElse(nodeParams.routerConf.searchMaxFeePct)
+    val maxFeeBase = request.routeParams.map(_.maxFeeBase).getOrElse(nodeParams.routerConf.searchMaxFeeBase.toMilliSatoshi)
+    val channels = if (randomize) {
+      Random.shuffle(balances.filter(p => p.remoteNodeId != request.targetNodeId))
+    } else {
+      balances.filter(p => p.remoteNodeId != request.targetNodeId).sortBy(_.canSend)
+    }
+    channels.foldLeft((remaining, localPayments)) {
+      case ((MilliSatoshi(0), payments), _) => (0 msat, payments)
+      case ((remaining, payments), channel) =>
+        require(remaining > 0.msat, s"payment splitting error: remaining amount must not be negative ($remaining): sending $toSend to ${request.targetNodeId} with balances=$balances, channels=$channels network=${networkStats.capacity}, fees=($maxFeeBase, $maxFeePct)")
+        // We use network statistics with a random factor to decide on the maximum amount for child payments.
+        // The current choice of parameters is completely arbitrary and could be made configurable.
+        // We could also learn from previous payment failures to dynamically tweak that value.
+        val maxAmount = networkStats.capacity.percentile75.toMilliSatoshi * ((75.0 + Random.nextInt(25)) / 100)
+        if (remaining <= maxAmount) {
+          val childAmount = Seq(remaining, channel.canSend * (1 - maxFeePct), channel.canSend - maxFeeBase).min
+          if (childAmount > 0.msat) {
+            val childPayment = createChildPayment(nodeParams, request, childAmount, channel)
+            (remaining - childPayment.finalPayload.amount, childPayment +: payments)
+          } else {
+            (remaining, payments)
+          }
+        } else {
+          val childCount = math.ceil(channel.canSend.min(remaining).toLong.toDouble / maxAmount.toLong).toInt
+          // We can't use all the available balance because we need to take fees into account and we don't know the
+          // exact fee before-hand because we don't know the rest of the route yet.
+          // Splitting into multiple HTLCs in the same channel will also increase the size of the CommitTx (and thus its
+          // fee), which decreases the available balance.
+          // We need to take that into account when trying to send multiple payments through the same channel.
+          val htlcsFees = Transactions.htlcOutputFee(nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(nodeParams.onChainFeeConf.feeTargets.commitmentBlockTarget)) * childCount
```

Comment:
That's true. That means we need to change a bit what's exposed in UsableBalances, I'll see how impactful that is.

Reply:
I'm deep down into changing this and it changes the `splitPayment` method a bit. I'm having some troubles because creating test `Commitments` objects is a bit painful, I'm trying to simplify it.
Please hold off reviewing the `splitPayment` method until I update it to use `Commitments`.